File: quick_sort_step.java
import java.util.*;
import java.io.*;

public class quick_sort_step {

    public static class Pair {
        int i;         // dataset integer
        String s;      // dataset string

        Pair(int i, String s) {
            this.i = i;
            this.s = s;
        }

        @Override
        public String toString() {
            return i + "/" + s;
        }
    }

    public static ArrayList<String> sl = new ArrayList<>();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        final String fn = "dataset_sample_1000.csv";

        System.out.print("Enter start row: ");
        int sr = sc.nextInt();

        System.out.print("Enter end row: ");
        int er = sc.nextInt();

        ArrayList<Pair> d = readCSVRange(fn, sr, er);
        if (d.isEmpty()) {
            System.out.println("No data found.");
            sc.close();
            return;
        }

        // Add initial unsorted list to step log
        sl.add(formatList(d));

        // Apply quick sort with tail recursion optimization
        quickSort(d, 0, d.size() - 1);

        // Write to output file
        String outFile = "quick_sort_step_" + sr + "_" + er + ".txt";
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(outFile))) {
            for (String line : sl) {
                bw.write(line);
                bw.newLine();
            }
            System.out.println("Steps written to " + outFile);
        } catch (IOException e) {
            System.out.println("Error writing output: " + e.getMessage());
        }

        sc.close();
    }

    public static ArrayList<Pair> readCSVRange(String fn, int sr, int er) {
        ArrayList<Pair> list = new ArrayList<Pair>();
        int cr = 1;

        try (BufferedReader br = new BufferedReader(new FileReader(fn))) {
            String l;
            while ((l = br.readLine()) != null) {
                if (cr >= sr && cr <= er) {
                    String[] parts = l.split(",", 2);
                    if (parts.length == 2) {
                        int number = Integer.parseInt(parts[0].trim());
                        String text = parts[1].trim();
                        list.add(new Pair(number, text));
                    }
                }
                cr++;
                if (cr > er) break;
            }
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
        }

        return list;
    }

    public static void quickSort(ArrayList<Pair> list, int left, int right) {
    if (left < right) {
        int pi = partition(list, left, right);
        sl.add("pi=" + pi + " " + formatList(list));
        quickSort(list, left, pi - 1);
        quickSort(list, pi + 1, right);
    }
}

    public static int partition(ArrayList<Pair> list, int low, int high) {
        Pair pivot = list.get(high);
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (list.get(j).i <= pivot.i) {
                Collections.swap(list, ++i, j);
            }
        }
        Collections.swap(list, i + 1, high);
        return i + 1;
    }

    public static String formatList(ArrayList<Pair> list) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < list.size(); i++) {
            sb.append(list.get(i));
            if (i < list.size() - 1) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }
}

File: quick_sort_step.cpp
#include <bits/stdc++.h>
using namespace std;

// Global start and end row
int sr, er;

// Output file stream
ofstream outfile;

// Print array from sr to er to file
void print_array(const vector<pair<int, string>>& vec) {
    outfile << "[";
    for (int i = sr; i <= er; ++i) {
        if (i != sr) outfile << ", ";
        outfile << vec[i].first << "/" << vec[i].second;
    }
    outfile << "]" << endl;
}

// Partition function
int partition(vector<pair<int, string>>& vec, int low, int high) {
    int pivot = vec[high].first;
    int i = (low - 1);
    for (int j = low; j <= high - 1; j++) {
        if (vec[j].first <= pivot) {
            i++;
            swap(vec[i], vec[j]);
        }
    }
    swap(vec[i + 1], vec[high]);
    return i + 1;
}

// Quick sort function (unchanged signature)
void quick_sort_step(vector<pair<int, string>>& vec, int low, int high) {
    if (low < high) {
        int pi = partition(vec, low, high);
        outfile << "pi=" << pi << " ";
        print_array(vec);
        quick_sort_step(vec, low, pi - 1);
        quick_sort_step(vec, pi + 1, high);
    }
}

int main() {
    ifstream file("dataset_sample_1000.csv");
    if (!file.is_open()) {
        cerr << "Error opening file!" << endl;
        return 1;
    }

    string line;
    vector<pair<int, string>> data;

    cout << "Enter start row and end row (0-based index): ";
    while (true) {
        cin >> sr >> er;
        if (sr >= 0 && er >= sr) break;
        cout << "Invalid input. Please enter valid start and end rows: ";
    }

    // Parse CSV
    while (getline(file, line)) {
        stringstream ss(line);
        string num_str, str_val;
        if (getline(ss, num_str, ',') && getline(ss, str_val)) {
            try {
                int num = stoi(num_str);
                data.push_back({num, str_val});
            } catch (...) {}
        }
    }
    file.close();

   
    string filename = "quick_sort_step_" + to_string(sr) + "_" + to_string(er) + ".txt";
    outfile.open(filename);
    if (!outfile.is_open()) {
        cerr << "Error opening output file!" << endl;
        return 1;
    }

  
    print_array(data);

  
    quick_sort_step(data, sr, er);

    outfile.close();
    return 0;
}
